File: Visual_Basic/Lists/SkipList/SkipList.vb
'******************************************************
' *  SkipList.vb
' *  Created by Stephen Hall on 11/20/17.
' *  Copyright (c) 2017 Stephen Hall. All rights reserved.
' *  Skip List implementation in Visual Basic
' *******************************************************

Namespace Lists.SkipList
	''' <summary>
	''' Skip list class
	''' </summary>
	''' <typeparam name="T">Generic type</typeparam>
	Public Class SkipList(Of T As IComparable)
		''' <summary>
		''' Node for Skip List class
		''' </summary>
		Public Class Node
			''' <summary>
			''' public members of the node class
			''' </summary>
			Public Data As T
			Public NodeList As List(Of Node)

			''' <summary>
			''' Node Constructor
			''' </summary>
			''' <param name="data1">data for the node to hold</param>
			Public Sub New(data1 As T)
				Data = data1
				NodeList = New List(Of Node)()
			End Sub

			''' <summary>
			''' Gets the level of the Node
			''' </summary>
			''' <returns>level of the node</returns>
			Public Function Level() As Integer
				Return NodeList.Count - 1
			End Function
		End Class

		''' <summary>
		''' Private members of the Skip list class
		''' </summary>
		Private _head As Node
		Private _max As Integer
		Private _size As Integer
		Private Const PROBABILITY As Double = 0.5
		Private ReadOnly _rnd As Random

		''' <summary>
		''' Skip List Constructor
		''' </summary>
		Public Sub New()
			_size = 0
			_max = 0
			_rnd = New Random(DateTime.Now.Second)
			_head = New Node(Nothing)
			' a Node with value null marks the beginning
				' null marks the end
			_head.NodeList.Add(Nothing)
		End Sub

		''' <summary>
		''' Adds a node into the Skip List
		''' </summary>
		''' <param name="data">data to add into the list</param>
		''' <returns>success|fail</returns>
		Public Function Add(data As T) As Boolean
			If Contains(data) Then
				Return False
			End If

			_size += 1
			Dim level = 0
			' random number from 0 to max + 1 (inclusive)
			While _rnd.NextDouble() < PROBABILITY
				level += 1
			End While
			While level > _max
				' should only happen once
				_head.NodeList.Add(Nothing)
				_max += 1
			End While

			Dim node As New Node(data)
			Dim current As Node = _head

			Do
				current = FindNext(data, current, level)
				node.NodeList(0) = current.NodeList(level)
				current.NodeList(level) = node
			Loop While Math.Max(Threading.Interlocked.Decrement(level),level + 1) > 0
			Return True
		End Function

		''' <summary>
		''' Finds a node in the list with the same data
		''' </summary>
		''' <param name="data">data to find</param>
		''' <returns>Node found</returns>
		Public Function Find(data As T) As Node
			Return Find(data, _head, _max)
		End Function

		''' <summary>
		''' Returns node with the greatest value
		''' </summary>
		''' <param name="data">data to find</param>
		''' <param name="current">current Node</param>
		''' <param name="level">level to start form</param>
		''' <returns>current node</returns>
		Private Function Find(data As T, current As Node, level As Integer) As Node
			Do
				current = FindNext(data, current, level)
			Loop While Math.Max(Threading.Interlocked.Decrement(level),level + 1) > 0
			Return current
		End Function

		''' <summary>
		''' Returns the node at a given level with highest value less than data
		''' </summary>
		''' <param name="data">data to find</param>
		''' <param name="current">current node</param>
		''' <param name="level">current level</param>
		''' <returns>highest node</returns>
		Private Function FindNext(data As T, current As Node, level As Integer) As Node
			Dim [next] As Node = current.NodeList(level)

			While [next] IsNot Nothing
				Dim value As T = [next].Data
				If LessThan(data, value) Then
					Exit While
				End If
				current = [next]
				[next] = current.NodeList(level)
			End While
			Return current
		End Function

		''' <summary>
		''' gets the size of the list
		''' </summary>
		''' <returns>size of the list</returns>
		Public Function Size() As Integer
			Return _size
		End Function

		''' <summary>
		''' Determines if the object is in the list or not
		''' </summary>
		''' <param name="data">object to test</param>
		''' <returns>true|false</returns>
		Public Function Contains(data As T) As Boolean
			Dim node As Node = Find(data)
			Return node IsNot Nothing AndAlso Not node.Data.Equals(Nothing) AndAlso EqualTo(node.Data, data)
		End Function

		''' <summary>
		''' Determines if a is less than b
		''' </summary>
		''' <param name="a">generic type to test</param>
		''' <param name="b">generic type to test</param>
		''' <returns>true|false</returns>
		Private Shared Function LessThan(a As T, b As T) As Boolean
			Return a.CompareTo(b) < 0
		End Function

		''' <summary>
		''' Determines if a is equal to b
		''' </summary>
		''' <param name="a">generic type to test</param>
		''' <param name="b">generic type to test</param>
		''' <returns>true|false</returns>
		Private Shared Function EqualTo(a As T, b As T) As Boolean
			Return a.CompareTo(b) = 0
		End Function
	End Class
End Namespace

File: C++/Lists/SkipList/SkipList.cpp
/*******************************************************
 *  @file SkipList.cpp
 *  @author Stephen Hall
 *  @date 12/04/17.
 *  @copyright 2017 Stephen Hall. All rights reserved.
 *  @details Skip List implementation in C++
 ********************************************************/

using namespace Lists::SkipList;
using SkipList::Node;
#include "SkipList.h"

/**
 * Node Default constructor
 * @tparam T: Generic type
 */
template<typename T>
Node::Node()
{
	data = NULL;
	nodeList = new std::vector<T>();
}

/**
 * Node constructor with value
 * @tparam T: Generic Type
 * @param data: Data of the node
 */
template<typename T>
Node::Node(T data)
{
	this->data = data;
	nodeList = new std::vector<T>();
}

/**
 * Node Destructor
 */
Node::~Node()
{
	for(int i = 0; i < nodeList->size(); ++i)
		delete nodeList[i];
	nodeList->clear();
	delete nodeList;
}

/**
 * Gets the node level
 * @return int: Level of the node
 */
int Node::level()
{
	return nodeList->size() - 1;
}

/**
 * Skip List constructor
 */
SkipList::SkipList()
{
	size = 0;
	max = 0;
	head = new Node(NULL);
	// a Node with value null marks the beginning
	head->nodeList->push_back(NULL);
}

/**
 * Skip List destructor
 */
SkipList::~SkipList()
{
	destroy();
}

/**
 * Adds an item into the List
 * @tparam T: Generic type
 * @param data: Data to add into the list
 * @return bool: success|fail
 */
template<typename T>
bool SkipList::add(T data)
{
	if(contains(data))
		return false;

	size++;
	int level = 0;
	// random number from 0 to max + 1 (inclusive)
	while (rand() < PROBABILITY)
		level++;

	while(level > max)
	{
		// should only happen once
		head->nodeList->push_back(NULL);
		max++;
	}

	Node* node = new Node(data);
	Node* current = head;

	do
	{
		current = findNext(data, current, level);
		node->nodeList->push_back(0, current->nodeList[level]);
		current->nodeList->at(level) = node;
	} while ((level--) > 0);
	return true;
}

/**
 * Finds a node in the list with the same data
 * @tparam T: Generic type
 * @param data: data to find
 * @return Node: Node found or NULL
 */
template<typename T>
Node* SkipList::find(T data)
{
	return find(data, head, max);
}

/**
 * Gets the size of the List
 * @return int: size of the list
 */
int SkipList::length()
{
	return size;
}

/**
 * Determines if the object is in the list or not
 * @tparam T: Generic type
 * @param data: object to test
 * @return boolean: true|false
 */
template<typename T>
bool SkipList::contains(T data)
{
	Node* node = find(data);
	return ((node != NULL) && (node->data != NULL) && (node->data == data));
}

/**
 * Destroys the skip list
 */
void SkipList::destroy()
{
	delete head;
}

/**
 * Returns the node at a given level with highest value less than data
 * @tparam T: Generic Type
 * @param data: data to find
 * @param current: current node
 * @param level: current level
 * @return Node: highest node
 */
template<typename T>
Node* SkipList::findNext(T data, Node* current, int level)
{
	Node* next = current->nodeList->at(level);

	while(next != NULL)
	{
		T value = next->data;
		if(data < value)
			break;
		current = next;
		next = current->nodeList->at(level);
	}
	return current;
}

/**
 * Returns node with the greatest value
 * @tparam T: Generic type
 * @param data: data to find
 * @param current: current Node
 * @param level: level to start form
 * @return Node: current node
 */
template<typename T>
Node* SkipList::find(T data, Node* current, int level)
{
	do
	{
		current = findNext(data, current, level);
	} while((level--) > 0);
	return current;
}
